src/core/io/src/4C_io_input_parameter_container.hpp

```diff
@@ -119,118 +70,169 @@ namespace Core::IO
-        return access(mapdata_, name);
-      else if constexpr (std::is_same_v<T, std::string>)
-        return access(stringdata_, name);
-      else
-        return access_any(anydata_, name);
-    }
+    const T* get_if(const std::string& name) const;
 
+    /**
+     * Print the data in the container to the output stream @p os.
+     */
+    void print(std::ostream& os) const;
 
    private:
-    //! a map to store integer data in
-    std::map<std::string, int> intdata_;
+    //! Entry stored in the container.
+    struct Entry
+    {
+      //! The actual data.
+      std::any data;
 
-    //! a map to store double data in
-    std::map<std::string, double> doubledata_;
+      //! A function to print the data.
+      std::function<void(std::ostream&, const std::any&)> print;
+    };
 
-    //! a map to store bool data in
-    std::map<std::string, bool> booldata_;
+    //! Data stored in this container.
+    std::map<std::string, Entry> entries_;
 
-    //! a map to store vector integer data in
-    std::map<std::string, std::vector<int>> vecintdata_;
+    //! Groups present in this container. Groups are InputParameterContainers themselves.
+    std::map<std::string, InputParameterContainer> groups_;
+  };  // class InputParameterContainer
 
-    //! a map to store vector double data in
-    std::map<std::string, std::vector<double>> vecdoubledata_;
+}  // namespace Core::IO
 
-    //! a map to store maps of stl vector data
-    std::map<std::string, std::map<int, std::vector<double>>> mapdata_;
 
-    //! a map to store string data in
-    std::map<std::string, std::string> stringdata_;
+// --- template and inline functions ---//
 
-    //! a map to store anything
-    std::map<std::string, std::any> anydata_;
 
-    //! Groups present in this container. Groups are InputParameterContainers themselves.
-    std::map<std::string, InputParameterContainer> groups_;
-  };  // class InputParameterContainer
-}  // namespace Core::IO
+namespace Core::IO::Internal::InputParameterContainerImplementation
+{
+  template <typename T>
+  const T* try_get_any_data(const std::string& name, const std::any& data)
+  {
+    if (typeid(T) == data.type())
+    {
+      const T* any_ptr = std::any_cast<T>(&data);
+      FOUR_C_ASSERT(any_ptr != nullptr, "Implementation error.");
+      return any_ptr;
+    }
+    else
+    {
+      FOUR_C_THROW(
+          "You tried to get the data named %s from the container as type '%s'.\n"
+          "Actually, it has type '%s'.",
+          name.c_str(), Core::Utils::get_type_name<T>().c_str(),
+          Core::Utils::try_demangle(data.type().name()).c_str());
+    }
+  }
 
-// << operator
-std::ostream& operator<<(std::ostream& os, const Core::IO::InputParameterContainer& cont);
+  // Default printer if not printable.
+  template <typename T>
+  struct PrintHelper
+  {
+    void operator()(std::ostream& os, const std::any& data) { os << "<not printable> "; }
```

Comment:
```suggestion
    void operator()(std::ostream& os, const std::any& data) { os << "<not printable>"; }
```

Why is there a space at the end? Applies also to the other cases

Reply:
It has been like this before. The print function of the container is not very nice and just dumps everything into a line. Could be improved in the future.